src/lib.rs

```diff
@@ -126,6 +134,14 @@ where
             attempt += 1;
         }
     }
+    /// Sets the function to be called before each retry; it will not be called before the first execution.
+    pub fn on_retry<F>(&mut self, on_retry: F) -> &mut Self
+    where
+        F: Fn() + Send + Sync + 'static,
+    {
+        self.on_retry = Some(Box::new(on_retry));
+        self
+    }
```

Comment:
Can you make it so that the function takes a tuple `|(res: &Result<T, E>, attempt: 32)|`?

Also worth considering is that in the next [branch I am merging](https://github.com/theelderbeever/mulligan/tree/1-create-trait-jitter-and-trait-backoff-to-improve-extensibility) a lot of stuff will transition to generics where the static type of Mulligan mutates with each setter. No need to implement those. I can make the changes as well.

Reply:
I have modified the function signature. Below is a simple example to clarify the purpose of the parameters:

```rust
mulligan::until_ok()
            .stop_after(3)
            .full_jitter()
            .fixed(Duration::from_millis(200))
            .on_retry(|res, attempt| { println!("[on_retry] start to call retry() again. In last attempt = {}, result = {:?}", attempt, res) })
            .retry(|| async { this_errors("[retry] call .retry() and failed").await })
            .await
```

The output is:
```text
[retry] call `.retry()` and failed
[on_retry] start to call retry() again. In last attempt = 0, result = Err(Custom { kind: Other, error: "uh oh!" })
[retry] call `.retry()` and failed
[on_retry] start to call retry() again. In last attempt = 1, result = Err(Custom { kind: Other, error: "uh oh!" })
[retry] call `.retry()` and failed
[on_retry] start to call retry() again. In last attempt = 2, result = Err(Custom { kind: Other, error: "uh oh!" })
[retry] call `.retry()` and failed
```

Additionally, for your reference, I found that when it is set to attempt 3 times, it actually executes a maximum of 4 times. I'm not sure if this is a feature or a bug.